src/coreclr/jit/helperexpansion.cpp

```diff
@@ -1821,171 +1821,378 @@ PhaseStatus Compiler::fgLateCastExpansion()
-//    stmt  - Statement containing the cast helper
-//    call  - The cast helper
+//    comp               - Compiler instance
+//    castHelper         - Cast helper call to expand
+//    commonCls          - [out] Common denominator class for the fast and the fallback paths.
+//    likelihood         - [out] Likelihood of successful type check (0..100)
+//    typeCheckFailed    - [out] Action to perform if the type check fails
+//    typeCheckPassed    - [out] Action to perform if the type check passes
 //
 // Returns:
-//    True if expanded, false otherwise.
+//    Likely class handle or NO_CLASS_HANDLE
 //
-bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt, GenTreeCall* call)
+static CORINFO_CLASS_HANDLE PickCandidateForTypeCheck(Compiler*              comp,
+                                                      GenTreeCall*           castHelper,
+                                                      CORINFO_CLASS_HANDLE*  commonCls,
+                                                      unsigned*              likelihood,
+                                                      TypeCheckFailedAction* typeCheckFailed,
+                                                      TypeCheckPassedAction* typeCheckPassed)
 {
-    if (!call->IsHelperCall())
-    {
-        return false;
-    }
-
-    if ((call->gtCallMoreFlags & GTF_CALL_M_CAST_CAN_BE_EXPANDED) == 0)
+    if (!castHelper->IsHelperCall() || ((castHelper->gtCallMoreFlags & GTF_CALL_M_CAST_CAN_BE_EXPANDED) == 0))
     {
         // It's not eligible for expansion (already expanded in importer)
         // To be removed once we move cast expansion here completely.
-        return false;
+        return NO_CLASS_HANDLE;
     }
 
-    bool isInstanceOf = false;
-    switch (call->GetHelperNum())
+    // Helper calls are never tail calls
+    assert(!castHelper->IsTailCall());
+
+    // is it "castclass" or "isinst"?
+    bool isCastClass;
+
+    const unsigned helper = castHelper->GetHelperNum();
+    switch (helper)
     {
-        case CORINFO_HELP_ISINSTANCEOFINTERFACE:
+        case CORINFO_HELP_CHKCASTARRAY:
+        case CORINFO_HELP_CHKCASTANY:
+        case CORINFO_HELP_CHKCASTINTERFACE:
+        case CORINFO_HELP_CHKCASTCLASS:
+            isCastClass = true;
+            break;
+
         case CORINFO_HELP_ISINSTANCEOFARRAY:
         case CORINFO_HELP_ISINSTANCEOFCLASS:
         case CORINFO_HELP_ISINSTANCEOFANY:
-            isInstanceOf = true;
+        case CORINFO_HELP_ISINSTANCEOFINTERFACE:
+            isCastClass = false;
             break;
 
-        case CORINFO_HELP_CHKCASTINTERFACE:
-        case CORINFO_HELP_CHKCASTARRAY:
-        case CORINFO_HELP_CHKCASTCLASS:
-        case CORINFO_HELP_CHKCASTANY:
-            break;
+        // These are never expanded:
+        // CORINFO_HELP_ISINSTANCEOF_EXCEPTION
+        // CORINFO_HELP_CHKCASTCLASS_SPECIAL
+        // CORINFO_HELP_READYTORUN_ISINSTANCEOF,
+        // CORINFO_HELP_READYTORUN_CHKCAST,
```

Comment:
Just for my curiosity, why not expand the R2R version if we have static PGO?

Reply:
Good point, but importer also always gives up on them currently, but I'll add it to my todo